### PR TITLE
Fixes to dynamic rendering and occlusion queries.

### PR DESCRIPTION
**Pull request by billhollings**
This PR fixes some issues with dynamic rendering and occlusion queries, and modifies the some Metal debug labelling to make it easier to distinguish attachment clearing activities within a GPU capture.

- Fix crash when `VkCommandBufferInheritanceInfo::renderPass` is `VK_NULL_HANDLE`.
- Do not clear attachments when dynamic rendering is resumed.
- Allow ending dynamic rendering to trigger next multiview pass if needed.
- Move deciding to begin next multiview pass to `MVKCommandEncoder`.
- Fix premature caching of occlusion query results during tessellation rendering. Tessellation ends Metal renderpass for compute control and eval stages. Wait until end of Metal renderpass after rasterization stage.
- `vkCmdCopyQueryPoolResults()`: Fix loss of queries when query count is not a multiple of GPU threadgroup execution width.
- Disable occlusion recording while clearing attachments or render area.
- `MVKCmdClearAttachments` improve labelling of `MTLDebugGroup` to better distinguish clearing renderpass render area from `vkCmdClearAttachments()` in an Xcode GPU capture (unrelated but helpful during debugging).
- `MVKCmdClearAttachments` re-order member variables to optimize memory requirements (unrelated).
- `MVKCommandBuffer` remove unused renderpass tracking functions (unrelated).

### CODE REVIEW

File: MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm [hunk 1]
```diff
@@ -183,9 +185,16 @@
...
 		cmdEncoder->setComputeBytes(mtlComputeCmdEnc, _availability.data(), _availability.size() * sizeof(Status), 5);
 		_availabilityLock.unlock();
+
 		// Run one thread per query. Try to fill up a subgroup.
-		[mtlComputeCmdEnc dispatchThreadgroups: MTLSizeMake(max(queryCount / mtlCopyResultsState.threadExecutionWidth, NSUInteger(1)), 1, 1)
-						  threadsPerThreadgroup: MTLSizeMake(min(NSUInteger(queryCount), mtlCopyResultsState.threadExecutionWidth), 1, 1)];
+		NSUInteger threadCount = NSUInteger(queryCount);
+		NSUInteger threadExecutionWidth = mtlCopyResultsState.threadExecutionWidth;
+		NSUInteger tgWidth = min(threadCount, threadExecutionWidth);
+		NSUInteger tgCount = threadCount / threadExecutionWidth;
```
Comment (cdavis5e) [inline MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:193]: What happens when there are fewer queries than will fill up an SIMD-group?

File: MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm [hunk 2]
```diff
@@ -281,12 +290,9 @@
 }
 
 void MVKOcclusionQueryPool::beginQueryAddedTo(uint32_t query, MVKCommandBuffer* cmdBuffer) {
+	// In multiview passes, one query is used for each view.
+	NSUInteger queryCount = cmdBuffer->getViewCount();
```
Comment (cdavis5e) [inline MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:294]: Wait, what happens when we're *not* in a secondary command buffer?

**Review by cdavis5e — CHANGES_REQUESTED** [review]
I had noticed some of these problems myself, while I was trying out a potential fix for some of the CTS tests.

File: MoltenVK/MoltenVK/Commands/MVKCommandBuffer.h [hunk 3]
```diff
@@ -188,7 +176,7 @@ class MVKCommandBuffer : public MVKDispatchableVulkanAPIObject,
 	MVKSmallVector<VkFormat, kMVKDefaultAttachmentCount> _colorAttachmentFormats;
 	MVKCommandPool* _commandPool;
 	VkCommandBufferInheritanceInfo _secondaryInheritanceInfo;
-	VkCommandBufferInheritanceRenderingInfo _inerhitanceRenderingInfo;
+	VkCommandBufferInheritanceRenderingInfo _secondaryInerhitanceRenderingInfo;
```
[the remark this answers is not in the document]
Comment (billhollings) [reply MoltenVK/MoltenVK/Commands/MVKCommandBuffer.h:179]: Thanks. Good eyes. Fixed.

File: MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm (hunk 1 above)
Comment (billhollings) [reply MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:193]: Is your concern about passing the right values to `dispatchThreadgroups:threadsPerThreadgroup:`?

If so, I think this

    NSUInteger threadCount = NSUInteger(queryCount);
    NSUInteger tgWidth = min(threadCount, threadExecutionWidth);

combined with  the guard code in the shader

    if (query >= numQueries) { return; }

 should cover that?

Or are you concerned that there is something deeper in the hardware that will run into trouble?

Comment (cdavis5e) [reply MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:193]: > Is your concern about passing the right values to `dispatchThreadgroups:threadsPerThreadgroup:`?

Yes.

What I'm trying to say is, if `queryCount` is less than `threadExecutionWidth`, then the division will come out zero, which isn't what we want. This division probably needs to round up.

Comment (billhollings) [reply MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:193]: > This division probably needs to round up.

Ah. Yes. The line following it does that.

Comment (cdavis5e) [reply MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:193]: How did I not notice that...?

Ah well. Never mind, then.

File: MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm (hunk 2 above)
Comment (billhollings) [reply MoltenVK/MoltenVK/GPUObjects/MVKQueryPool.mm:294]: It would default to 1. This was the same as existing behavior, since the existing `_lastMultiviewSubpass` was never set (`recordBeginRenderPass` was never called).

The `queryCount` value here is only used as a sanity check on multipass queries that approach the end of the visibility buffer, so I toyed with simply removing all this code.

However, in the end, I added a `MVKCurrentSubpassInfo` helper in `MVKCommandBuffer` to track the current renderpass, subpass, or dynamic rendering, to derive a `viewMask` for the command corresponding to the currently added subpass, replacing and expanding on what `_lastMultiviewSubpass` was supposed to do.

So, even though it's just doing an edge-case validation test, we've got some plumbing to track other subpass command info if we need to.

**Review by cdavis5e — APPROVED** [review]
